src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs

```diff
@@ -796,76 +842,127 @@ internal DefinitionResult GetDefinition(TextDocumentPosition textDocumentPositio
                         Locations = null
                     };
                 }
-                finally
+            }
+
+            // then check the parents 
+            foreach (var i in selectedToken.Item2.ToList())
+            {
+                var token = selectedToken.Item2.Dequeue();
+                // Strip "[" and "]"(if present) from the token text to enable matching with the suggestions.
+                // The suggestion title does not contain any sql punctuation
+                string tokenText = TextUtilities.RemoveSquareBracketSyntax(token.Text);
+                textDocumentPosition.Position.Line = token.StartLocation.LineNumber;
+                textDocumentPosition.Position.Character = token.StartLocation.ColumnNumber;
+                if (scriptParseInfo.IsConnected && Monitor.TryEnter(scriptParseInfo.BuildingMetadataLock))
+                {
+                    try
+                    {                    
+                        var result = QueueTask(textDocumentPosition, scriptParseInfo, connInfo, scriptFile, tokenText);
+                        lastResult = result;
+                        if (!result.IsErrorResult)
+                        {
+                            return result;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // if any exceptions are raised return error result with message
+                        Logger.Write(LogLevel.Error, "Exception in GetDefinition " + ex.ToString());
+                        return new DefinitionResult
+                        {
+                            IsErrorResult = true,
+                            Message = SR.PeekDefinitionError(ex.Message),
+                            Locations = null
+                        };
+                    }
+                    finally
+                    {
+                        Monitor.Exit(scriptParseInfo.BuildingMetadataLock);
+                    }
+                }
+                else
                 {
-                    Monitor.Exit(scriptParseInfo.BuildingMetadataLock);
+                    // User is not connected.
+                    return new DefinitionResult
+                    {
+                        IsErrorResult = true,
+                        Message = SR.PeekDefinitionNotConnectedError,
+                        Locations = null
+                    };
                 }
             }
-            else
+            return (lastResult != null) ? lastResult : null;
+        }
+   
+        /// <summary>
+        /// Wrapper around find token method 
+        /// </summary>
+        /// <param name="scriptParseInfo"></param>
+        /// <param name="startLine"></param>
+        /// <param name="startColumn"></param>
+        /// <returns> token index</returns>
+        private int FindTokenWrapper(ScriptParseInfo scriptParseInfo, int startLine, int startColumn)
```

Comment:
Minor: should this be a global utility? Does any other scenario need this? Also I'd rename to "FindTokenWithCorrectOffset" or similar. "Wrapper" doesn't make clear why this is.